File: ecmascript/bracket-push/bracket-push.js
const OPENING = ['(', '[', '{'];
const CLOSING = [')', ']', '}'];

const bracket = ([...str]) => {
  const stack = [];
  for (let i = 0; i < str.length; i++) {
    let b = str[i];
    if (OPENING.indexOf(b) >= 0) {
      stack.push(b);
    } else {
      if (OPENING.indexOf(stack.pop()) != CLOSING.indexOf(b))
        return false;
    }
  }
  return stack.length == 0 ? true : false;
};

export default bracket;

File: elixir/bracket-push/bracket_push.exs
defmodule BracketPush do
  @opening ["(", "[", "{"]
  @closing [")", "]", "}"]
  @brackets Enum.zip(@opening, @closing) |> Enum.into(%{})

  @doc """
  Checks that all the brackets and braces in the string are matched correctly, and nested correctly
  """
  @spec check_brackets(String.t) :: boolean
  def check_brackets(str) do
    str
    |> filter_brackets
    |> balance_brackets([])
  end

  defp filter_brackets(str) do
    str
    |> String.replace(~r/[^\(\)\[\]\{\}]/, "")
    |> String.codepoints
  end

  defp balance_brackets([], []), do: true
  defp balance_brackets([], _stack), do: false
  defp balance_brackets([b|brackets], stack) when b in @opening do
    balance_brackets brackets, [b|stack]
  end
  defp balance_brackets([b|brackets], [top|stack]) do
    case b == @brackets[top] do
      true ->
        balance_brackets brackets, stack
      _ ->
        false
    end
  end
  defp balance_brackets(_, _), do: false
end
